compiler/AST/type.cpp

```diff
@@ -613,6 +617,396 @@ void EnumType::accept(AstVisitor* visitor) {
+    case INTENT_CONST_REF: return "CR";
+    case INTENT_REF_MAYBE_CONST: return "RMC";
+    case INTENT_PARAM: return "P";
+    case INTENT_TYPE: return "T";
+    case INTENT_BLANK: return "";
+  }
+  return nullptr;
+}
+
+const char* FunctionType::retTagMnemonicMangled(RetTag tag) {
+  switch (tag) {
+    case RET_VALUE: return "";
+    case RET_REF: return "R";
+    case RET_CONST_REF: return "CR";
+    case RET_PARAM: return "P";
+    case RET_TYPE: return "T";
+  }
+  return nullptr;
+}
+
+const char* FunctionType::toStringMangledForCodegen() const {
+  std::ostringstream oss;
+
+  oss << "chpl_" << kindToString(kind_) << "_";
+
+  for (int i = 0; i < numFormals(); i++) {
+    auto f = this->formal(i);
+    bool skip = isIntentSameAsDefault(f->intent, f->type);
+    if (!skip) oss << intentTagMnemonicMangled(f->intent);
+    oss << typeToString(f->type) << "_";
+    if (f->name) oss << f->name;
+    oss << "_";
+  }
+
+  oss << "_";
+  if (returnIntent_ != RET_VALUE) {
+    oss << retTagMnemonicMangled(returnIntent_) << "_";
+  }
+
+  oss << typeToString(returnType_);
+  if (throws_) oss << "_throws";
+
+  auto ret = astr(oss.str());
+  return ret;
+}
+
+size_t FunctionType::hash() const {
+  size_t ret = ((size_t) kind_);
+  std::hash<void*> hasherPtr;
+  std::hash<const char*> hasherConstCharPtr;
+  std::hash<bool> hasherBool;
+
+  // I think it's fine to hash the pointers here because types don't really
+  // have a meaningful way to distinguish on contents, and should be unique,
+  // while the formal names are all canonical using 'astr'.
+  for (auto& formal : formals_) {
+    ret = chpl::hash_combine(ret, hasherPtr(formal.type));
+    ret = chpl::hash_combine(ret, ((size_t) formal.intent));
+    ret = chpl::hash_combine(ret, hasherConstCharPtr(formal.name));
+  }
+
+  ret = chpl::hash_combine(ret, ((size_t) returnIntent_));
+  ret = chpl::hash_combine(ret, hasherPtr(returnType_));
+  ret = chpl::hash_combine(ret, hasherBool(throws_));
+  ret = chpl::hash_combine(ret, hasherBool(isAnyFormalNamed_));
+
+  return ret;
+}
+
+bool
+FunctionType::Formal::operator==(const FunctionType::Formal& rhs) const {
+  return this->type == rhs.type &&
+    this->intent == rhs.intent &&
+    this->name == rhs.name;
+}
+
+bool FunctionType::equals(const FunctionType* rhs) const {
+  return this->kind_ == rhs->kind_ &&
+    this->formals_ == rhs->formals_ &&
+    this->returnIntent_ == rhs->returnIntent_ &&
+    this->returnType_ == rhs->returnType_ &&
+    this->throws_ == rhs->throws_ &&
+    this->isAnyFormalNamed_ == rhs->isAnyFormalNamed_ &&
+    this->userTypeString_ == rhs->userTypeString_;
+}
+
+bool FunctionType::isGeneric() const {
+  auto rt = returnType();
+  if (rt == dtUnknown || rt == dtAny ||
+      rt->symbol->hasFlag(FLAG_GENERIC)) {
+    return true;
+  }
+
+  for (int i = 0; i < numFormals(); i++) {
+    auto f = formal(i);
+    if (f->type == dtUnknown || f->type == dtAny ||
+        f->type->symbol->hasFlag(FLAG_GENERIC)) {
```

Comment:
1) is this really not a helper function somewhere already?
2) can this also be a method on the formal?

Reply:
I see `AggregateType::isGeneric()` but would have to go crawling to find something more general. I moved the formal check to `Formal`.